source/sorted_table_model.ts

```diff
@@ -0,0 +1,198 @@
+    const rowOrdering = [];
+    for(let i = 0; i < this.translatedTable.rowCount; ++i) {
+      rowOrdering.push(i);
+    }
+    rowOrdering.sort((a, b) => this.compareRows(a, b));
+    for(let i = 0; i < rowOrdering.length; ++i) {
+      this.translatedTable.moveRow(rowOrdering[i], i);
+      for(let j = i + 1; j < rowOrdering.length; ++j) {
+        if(rowOrdering[j] < rowOrdering[i]) {
+          ++rowOrdering[j];
+        }
+      }
+    }
+  }
+
+  private compareRows(left: number, right: number) {
+    for(let i = 0; i < this.sortPriority.length; ++i) {
+      const value = this.comparator.compareValues(
+        this.translatedTable.get(left, this.sortPriority[i]),
+        this.translatedTable.get(right, this.sortPriority[i]));
+      if(value !== 0) {
+        if(this.sortOrder[this.sortPriority[i]] === SortOrder.ASCENDING) {
+          return value;
+        } else {
+          return -value;
+        }
+      }
+    }
+    return 0;
+  }
+
+  private handleSourceOperation = (operation: Operation) => {
+    if(operation instanceof AddRowOperation) {
+      this.sourceAdd(operation);
+    } else if(operation instanceof MoveRowOperation) {
+      this.sourceMove(operation);
+    } else if(operation instanceof RemoveRowOperation) {
+      this.transactionLog.push(operation);
+    } else if(operation instanceof UpdateOperation) {
+      this.sourceUpdate(operation);
+    } else if(operation instanceof Transaction) {
+      this.beginTransaction();
+      operation.operations.forEach(this.handleSourceOperation);
+      this.endTransaction();
+    }
+  }
+
+  private sourceAdd(operation: AddRowOperation) {
+    const sortedIndex = this.findSortedIndex(operation.index);
+    ++this.movesToIgnore;
+    this.translatedTable.moveRow(operation.index, sortedIndex);
+    this.transactionLog.push(new AddRowOperation(sortedIndex));
+  }
+
+  private sourceMove(operation: MoveRowOperation) {
+    if(this.movesToIgnore > 0) {
+      --this.movesToIgnore;
+    } else if(operation.source !== operation.destination){
+      this.transactionLog.push(operation);
+    }
+  }
+
+  private sourceUpdate(operation: UpdateOperation) {
+    this.beginTransaction();
+    const sortedIndex = this.findSortedIndex(operation.row);
+    this.translatedTable.moveRow(operation.row, sortedIndex);
+    this.transactionLog.push(new UpdateOperation(
+      sortedIndex, operation.column));
+    this.endTransaction();
+  }
+
+  private findSortedIndex(source: number): number {
+    if(source !== 0 &&
+        this.compareRows(source - 1, source) > 0) {
+      return this.findInHead(0, source - 1, source);
+    } else if(source !== this.rowCount - 1 &&
+        this.compareRows(source, source + 1) > 0) {
+      return this.findInTail(source + 1, this.rowCount - 1, source);
+    } else {
+      return source;
+    }
+  }
+
+  private findInHead(start: number, end: number, indexOfValue: number) {
+    while(start < end) {
+      const middle = Math.floor((start + end) / 2);
```

Comment:
Fix indentation, also fix the bug that is both found here and at ```findInTail``` but reading about "Java binary search bug".